File: 1621-number-of-subsequences-that-satisfy-the-given-sum-condition/number-of-subsequences-that-satisfy-the-given-sum-condition.py
class Solution:
    def numSubseq(self, nums: List[int], target: int) -> int:
        M = 1e9 + 7
        nums.sort()
        n = len(nums)
        res = 0
        j = n-1
        power = [1] * n
        for i in range(1,n):
            power[i] = (power[i-1] * 2) % M
        for i in range(n):
            if nums[i] + nums[j] <= target:
                res += power[j-i]
                res %= M
            else:
                while i < j and nums[i] + nums[j] > target:
                    j -= 1
                if nums[i] + nums[j] <= target:
                    res += power[j-i]
                    res %= M
                else:
                    break
        return int(res)

File: 1621-number-of-subsequences-that-satisfy-the-given-sum-condition/number-of-subsequences-that-satisfy-the-given-sum-condition.cpp
class Solution {
public:
    int numSubseq(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        int M = 1e9 + 7, n = nums.size();
        vector<int> power(n,1);
        for (int i=1; i<n; ++i) power[i] = (power[i-1] * 2) % M;
        int i = 0, j = n-1, res = 0;
        while (i <= j) {
            if (nums[i] + nums[j] <= target) {
                res = (res + power[j-i] % M) % M;
                i++;
            } else j--;
        }
        return res;
    }
};